### PR TITLE
Graceful handling "connection: close" header

### PR DESCRIPTION
**Pull request by mspiewak**
Graceful handling "connection: close" header

### CODE REVIEW

File: lib/machine_gun/worker.ex [hunk 1]
```diff
@@ -304,4 +307,11 @@ defmodule MachineGun.Worker do
 
   def parse_reason({:shutdown, reason}), do: %Error{reason: reason}
   def parse_reason(reason), do: %Error{reason: reason}
+
+  defp is_last_request(headers) do
```
Comment (mponnuraj) [inline lib/machine_gun/worker.ex:311]: rename to `last_request?` ?

File: lib/machine_gun/worker.ex [hunk 2]
```diff
@@ -279,10 +282,10 @@ defmodule MachineGun.Worker do
          stream_ref,
          is_fin,
          from,
-         response,
+         %Response{headers: headers} = response,
          cancel_ref
        ) do
-    if is_fin == :fin do
+    if is_fin == :fin && !is_last_request(headers) do
       :ok = GenServer.reply(from, {:ok, response})
```
Comment (petrohi) [inline lib/machine_gun/worker.ex:289]: What does this mean in practice? We don't reply with `:ok` if the response contains connection close header?

**Review by petrohi — CHANGES_REQUESTED** [review]

File: lib/machine_gun/worker.ex (hunk 2 above)
Comment (mspiewak) [reply lib/machine_gun/worker.ex:289]: yes, if the response contains connection close header we will reply only after the connection is closed (line 261)

File: lib/machine_gun/worker.ex (hunk 1 above)
Comment (mspiewak) [reply lib/machine_gun/worker.ex]: done

File: lib/machine_gun/worker.ex (hunk 2 above)
Comment (petrohi) [reply lib/machine_gun/worker.ex:289]: Do we expect server initiate the closing after it returned connection close header?

Comment (mspiewak) [reply lib/machine_gun/worker.ex:289]: no, connection close header means that server wants to close the connection and client shouldn't be sending more requests on it. I traced packets with Wireshark and `FIN,ACK` frame is sent by a client side (gun) after receiving this header. after connection is down we receive :gun_down and :DOWN messages and the latter one resets `gun_pid` to nil. next time the worker is invoked new connection is gonna be created.

Comment (petrohi) [reply lib/machine_gun/worker.ex:289]: What if we only make this change:

```
if is_fin == :fin do
  :ok = GenServer.reply(from, {:ok, response})
  worker = clean_refs(worker, stream_ref, cancel_ref)
  if last_request?(headers) do
    :ok = :gun.close(gun_pid)      
    true = :erlang.demonitor(gun_ref, [:flush])
    %{worker | gun_pid: nil, gun_ref: nil}
  else
    worker
  end
  worker
end

Comment (mspiewak) [reply lib/machine_gun/worker.ex:286]: yes, it's much simpler. I've made suggested changes plus added clearing `streams` and `cancels` from the state after connection is terminated.

File: lib/machine_gun/worker.ex [hunk 3]
```diff
@@ -304,4 +309,21 @@ defmodule MachineGun.Worker do
...
+  end
+
+  defp close_connection(%Worker{gun_pid: gun_pid, gun_ref: gun_ref} = worker) do
+    :ok = :gun.close(gun_pid)
+    true = :erlang.demonitor(gun_ref, [:flush])
+
+    worker
+    |> Map.put(:gun_pid, nil)
+    |> Map.put(:gun_ref, nil)
+    |> reply_error(:normal)
```
Comment (petrohi) [inline lib/machine_gun/worker.ex:327]: Looks good! Let's use :closed as reason, also please update the structure with | to let the compile check its fields:

```
%{worker | gun_pid: nil, gun_ref: nil} |> reply_error(:closed)
```

**Review by petrohi — CHANGES_REQUESTED** [review]

**Review by petrohi — APPROVED** [review]

**Pull request merged**